src/rust/engine/graph/src/lib.rs

```diff
@@ -116,42 +165,71 @@ impl<N: Node> Entry<N> {
   }
 
   ///
-  /// Returns a Future for the Node's field.
+  /// Spawn the execution of the work on an Executor, which will cause it to execute
+  /// outside of the Graph lock, and allow us to call back into the graph lock to set the
+  /// final value.
   ///
-  fn get<C>(&mut self, context: &C, entry_id: EntryId) -> BoxFuture<N::Item, N::Error>
+  fn start<C>(
+    context: &C,
+    entry_key: &EntryKey<N>,
+    entry_id: EntryId,
+    run_token: RunToken,
+    generation: Generation,
+    previous_result: Option<Result<N::Item, N::Error>>,
+  ) -> EntryState<N>
   where
     C: NodeContext<Node = N>,
   {
-    let next_state = match &mut self.state {
-      &mut EntryState::NotStarted(run_token) => {
-        match &self.node {
-          &EntryKey::Valid(ref n) => {
-            let context2 = context.clone_for(entry_id);
-            let node = n.clone();
-
-            // Spawn the execution of the work on an Executor, which will cause it to execute
-            // outside of the Graph lock, and allow us to call back into the graph lock to set the
-            // final value.
-            context.spawn(future::lazy(move || {
-              let context = context2.clone();
-              node.run(context2).then(move |res| {
-                context.graph().complete(entry_id, run_token, res);
-                Ok(())
-              })
-            }));
-
-            EntryState::Running {
-              waiters: Vec::new(),
-              start_time: Instant::now(),
-              run_token,
-            }
-          }
-          &EntryKey::Cyclic(_) => EntryState::Completed {
-            result: Err(N::Error::cyclic()),
-            run_token: run_token,
-          },
+    // Increment the RunToken to uniquely identify this work.
+    let run_token = run_token.next();
+    match entry_key {
+      &EntryKey::Valid(ref n) => {
+        let context2 = context.clone_for(entry_id);
```

Comment:
We've generally been using `context2` to mean "clone of `context` which I was required to make because of annoying move semantics"; possibly call this `context_for_entry_id`?

Reply:
Fixed.